src/snek5000/make.py

```diff
@@ -128,3 +133,98 @@ def exec(self, *rules, dryrun=False, keep_incomplete=True, **kwargs):
+    change in the compiler configuration.
+
+    .. note::
+
+        There is a small caveat with this solution. If a series of jobs are
+        launched which uses various compiler configurations, then it would
+        rebuild Nek5000 tools and libraries again and again during
+        ``sim.output.post_init`` - and suddenly some library would be missing
+        during the ``sim.make.exec`` call. As long as all subsequent jobs use
+        the same compiler configuration, it should work.
+
+    """
+
+    def __init__(self):
+        #: ``NEK_SOURCE_ROOT`` is the working directory
+        self.path_run = Path(snek5000.get_nek_source_root())
+
+        #: The ``nek5000.smk`` Snakemake file from :mod:`snek5000.assets` is used here
+        self.file = snek5000.get_asset("nek5000.smk")
+
+        #: A file lock ``nek5000_make_config.yml.lock`` used to prevent race conditions
+        self.lock = FileLock(self.path_run / "nek5000_make_config.yml.lock")
+
+        #: A YAML file ``nek5000_make_config.yml`` to record compiler configuration
+        self.config_cache = self.path_run / "nek5000_make_config.yml"
+
+        #: List of files to build
+        self.targets = [
+            "bin/genbox",
+            "bin/genmap",
+            "3rd_party/gslib/lib/libgs.a",
+            "3rd_party/blasLapack/libblasLapack.a",
+        ]
+
+        # TODO: replace with Snek5000 log handler?
+        self.log_handler = []
+
+    def has_to_build(self, compiler_config):
+        compiler_config_str = yaml.safe_dump(compiler_config)
+        compiler_config_cache = (
+            self.config_cache.read_text() if self.config_cache.exists() else ""
+        )
+
+        if (
+            # One or more of the targets do not exist
+            any(not (self.path_run / target).exists() for target in self.targets)
+            or compiler_config_str != compiler_config_cache
+            # New compiler configuration
+        ):
+            # Save new configuration first
+            self.config_cache.write_text(compiler_config_str)
+
+            return True
+        else:
+            return False
+
+    def build(self, config):
+        """Build Nek5000 tools and third party libraries essential for a
+        simulation.
+
+        Parameters
+        ----------
+        config: dict
+            Snakemake configuration
+
+        Returns
+        -------
+        bool
+            ``True`` if workflow execution was successful.
+
+        """
+
+        compiler_config = {
+            key: config[key]
+            for key in (
+                "CC",
+                "FC",
+                "MPICC",
+                "MPIFC",
+                "CFLAGS",
+                "FFLAGS",
+            )
+        }
+
+        with self.lock:
+            # Only one process can inspect at a time. No timeout
+            if self.has_to_build(compiler_config):
+                return self.exec(*self.targets, config=config)
```

Comment:
We could clean nek5000 before the tests used for coverage to have this line covered.